lightly/models/simsiam.py

```diff
@@ -0,0 +1,159 @@
+from lightly.models.batchnorm import get_norm_layer
+from lightly.models._loader import _StateDictLoaderMixin
+
+
+def _prediction_mlp(in_dims: int = 2048, 
+                    h_dims: int = 512, 
+                    out_dims: int = 2048) -> nn.Sequential:
+    """Prediction MLP. The original paper's implementation has 2 layers, with 
+    BN applied to its hidden fc layers but no BN or ReLU on the output fc layer.
+
+    Note that the hidden dimensions should be smaller than the input/output 
+    dimensions (bottleneck structure). The default implementation using a 
+    ResNet50 backbone has an input dimension of 2048, hidden dimension of 512, 
+    and output dimension of 2048
+
+    Args:
+        in_dims:
+            Input dimension of the first linear layer.
+        h_dims: 
+            Hidden dimension of all the fully connected layers (should be a
+            bottleneck!)
+        out_dims: 
+            Output Dimension of the final linear layer.
+
+    Returns:
+        nn.Sequential:
+            The projection head.
+    """
+    l1 = nn.Sequential(nn.Linear(in_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l2 = nn.Linear(h_dims, out_dims)
+
+    prediction = nn.Sequential(l1, l2)
+    return prediction
+
+
+def _projection_mlp(in_dims: int = 2048, 
+                    h_dims: int = 2048, 
+                    out_dims: int = 2048) -> nn.Sequential:
+    """Projection MLP. The original paper's implementation has 3 layers, with 
+    BN applied to its hidden fc layers but no ReLU on the output fc layer.
+
+    Args:
+        in_dims:
+            Input dimension of the first linear layer.
+        h_dims: 
+            Hidden dimension of all the fully connected layers.
+        out_dims: 
+            Output Dimension of the final linear layer.
+
+    Returns:
+        nn.Sequential:
+            The projection head.
+    """
+    l1 = nn.Sequential(nn.Linear(in_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l2 = nn.Sequential(nn.Linear(h_dims, h_dims),
+                       nn.BatchNorm1d(h_dims),
+                       nn.ReLU(inplace=True))
+
+    l3 = nn.Sequential(nn.Linear(h_dims, out_dims),
+                       nn.BatchNorm1d(out_dims))
+
+    projection = nn.Sequential(l1, l2, l3)
+    return projection
+
+
+class SimSiam(nn.Module, _StateDictLoaderMixin):
+    """ Implementation of SimSiam network
+
+    Attributes:
+        backbone:
+            TODO
+        width:
+            Width of the ResNet.
+        num_ftrs:
+            Dimension of the embedding (before the projection head).
+        proj_hidden_dim:
+            TODO
+        pred_hidden_dim:
+            TODO
+        out_dim:
+            Dimension of the output (after the projection head).
+
+    """
+
+    def __init__(self,
+                 backbone: nn.Module,
```

Comment:
Please also rename `backbone` to `features` for consistency

Reply:
If you're refactoring the models would you be open to renaming the other names to backbone? I think it's more informative than features in this context

Reply:
On board with keeping it consistent though obviously

Reply:
Yes, we also like `backbone` better. In that case you can keep it the way it is and we will change the other names.